File: webapp/lib/formatters/base-formatter.ts
/**
 * Base formatter for all content types
 */

import { BrandProfile } from "../models/brand-profile";
import { EventContext } from "../models/event-context";

export interface GeneratedContent {
  [key: string]: string | string[];
}

export abstract class BaseFormatter {
  constructor(
    protected brandProfile: BrandProfile,
    protected eventContext?: EventContext
  ) {}

  abstract format(generatedContent: GeneratedContent): string;

  protected cleanOutput(text: string): string {
    let cleaned = text;
    for (const word of this.brandProfile.wordsToAvoid) {
      const pattern = new RegExp(word, "gi");
      cleaned = cleaned.replace(pattern, "");
    }
    // Clean up extra whitespace
    cleaned = cleaned.split(/\s+/).join(" ");
    return cleaned.trim();
  }

  protected validateLength(text: string, maxWords: number, fieldName: string): void {
    const wordCount = text.split(/\s+/).length;
    if (wordCount > maxWords) {
      throw new Error(
        `${fieldName} exceeds maximum ${maxWords} words (got ${wordCount} words)`
      );
    }
  }
}

File: nightlife_engine/formatters/base_formatter.py
"""Base formatter for all content types."""

from abc import ABC, abstractmethod
from typing import Dict, Any
from ..models.brand_profile import BrandProfile
from ..models.event_context import EventContext


class BaseFormatter(ABC):
    """Base class for all content formatters."""

    def __init__(self, brand_profile: BrandProfile, event_context: EventContext = None):
        """Initialize formatter with brand and event context."""
        self.brand_profile = brand_profile
        self.event_context = event_context

    @abstractmethod
    def format(self, generated_content: Dict[str, Any]) -> str:
        """
        Format the generated content into the final output.

        Args:
            generated_content: Dictionary containing generated content parts

        Returns:
            Formatted string ready for posting
        """
        pass

    def _clean_output(self, text: str) -> str:
        """Remove any forbidden words and clean the text."""
        cleaned = text
        for word in self.brand_profile.words_to_avoid:
            # Case-insensitive replacement
            import re
            pattern = re.compile(re.escape(word), re.IGNORECASE)
            cleaned = pattern.sub("", cleaned)

        # Clean up extra whitespace
        cleaned = " ".join(cleaned.split())
        return cleaned.strip()

    def _validate_length(self, text: str, max_words: int, field_name: str) -> None:
        """Validate text doesn't exceed maximum word count."""
        word_count = len(text.split())
        if word_count > max_words:
            raise ValueError(
                f"{field_name} exceeds maximum {max_words} words "
                f"(got {word_count} words)"
            )
